### PR TITLE
Fix `Scrollable.ensureVisible` throws an error in `DropdownMenu` inside nested scroll views

### LINKED ISSUE #146764: DropdownMenu throws exception when nested inside two SingleChildScrollView
### Steps to reproduce

1. Add a DropdownMenu inside two SingleChildScrollView
2. Run the debuger
3. Try to writte something on the field that matches one of the menu entries

### Expected results

It shouldn't throw an exception, the way it works if we delete the first SingleChildScrollView

### Actual results

It throws `renderer.parent != null is not true`

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: SingleChildScrollView(
        child: Card(
          child: SingleChildScrollView(
            child: DropdownMenu<String>(
                enableFilter: true,
                requestFocusOnTap: true,
                label: Text("test"),
                dropdownMenuEntries: [
                  DropdownMenuEntry(value: "1", label: "one"),
                  DropdownMenuEntry(value: "2", label: "two"),
                ]),
          ),
        ),
      ),
    );
  }
}

```

</details>

### Screenshots or Video

_No response_

### Logs

<details open><summary>Debug console output</summary>

```console
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: 'package:flutter/src/rendering/object.dart': Failed assertion: line 3345 pos 14: 'renderer.parent != null': is not true.
#0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:51:61)
#1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:40:5)
#2      RenderObject.getTransformTo (package:flutter/src/rendering/object.dart:3345:14)
#3      ScrollPosition.ensureVisible (package:flutter/src/widgets/scroll_position.dart:809:28)
#4      ScrollableState._performEnsureVisible (package:flutter/src/widgets/scrollable.dart:1041:55)
#5      Scrollable.ensureVisible (package:flutter/src/widgets/scrollable.dart:475:45)
#6      _DropdownMenuState.scrollToHighlight.<anonymous closure> (package:flutter/src/material/dropdown_menu.dart:450:20)
#7      SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1386:15)
#8      SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1322:11)
#9      SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:1169:5)
#10     _invoke (dart:ui/hooks.dart:312:13)
#11     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:399:5)
#12     _drawFrame (dart:ui/hooks.dart:283:31)
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel , 3.19.5, on Arch Linux 6.8.5-arch1-1, locale en_US.UTF-8)
    • Flutter version 3.19.5 on channel  at /usr/lib/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision archlinuxaur (unknown (arch linux aur package)), 2038-01-19
      03:14:08
    • Engine revision e76c95649884
    • Dart version 3.3.2
    • DevTools version 2.31.1

[✗] Android toolchain - develop for Android devices
    ✗ ANDROID_HOME = /opt/android-sdk
      but Android SDK not found at this location.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 17.0.6
    • cmake version 3.29.2
    • ninja version 1.11.1
    • pkg-config version 2.1.1

[✓] Android Studio (version 2023.1)
    • Android Studio at /opt/android-studio
    • Flutter plugin version 77.2.1
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Arch Linux 6.8.5-arch1-1
    • Chrome (web)    • chrome • web-javascript • Google Chrome 123.0.6312.122

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

</details>

### PR DESCRIPTION
**Pull request by TahaTesser**
fixes [DropdownMenu throws exception when nested inside nested SingleChildScrollView](https://github.com/flutter/flutter/issues/146764)

### Description 
`Scrollable.ensureVisible` in `DropdownMenu` inside nested scroll views after the `OverlayPortal` [migration](https://github.com/flutter/flutter/pull/130534) throws an error. 

The error is thrown by `getTransformTo` when `object.parent != null` is not true.  This PR fixes the error by getting `Scrollable` and call `ensureVisible`. 

### Code sample

<details>
<summary>expand to view the code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Material(
        child: SingleChildScrollView(
          child: Card(
            child: SingleChildScrollView(
              child: DropdownMenu<String>(
                label: Text('test'),
                menuHeight: 100,
                dropdownMenuEntries: <DropdownMenuEntry<String>>[
                  DropdownMenuEntry<String>(value: '1', label: 'one'),
                  DropdownMenuEntry<String>(value: '2', label: 'two'),
                  DropdownMenuEntry<String>(value: '3', label: 'three'),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}
```

</details>

### Before
Screenshot 2024-04-22 at 18 08 56

### After 
Screenshot 2024-04-22 at 18 09 10

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

File: packages/flutter/lib/src/material/dropdown_menu.dart [hunk 1]
```diff
@@ -524,7 +524,10 @@ class _DropdownMenuState<T> extends State<DropdownMenu<T>> {
     WidgetsBinding.instance.addPostFrameCallback((_) {
       final BuildContext? highlightContext = buttonItemKeys[currentHighlight!].currentContext;
       if (highlightContext != null) {
-        Scrollable.ensureVisible(highlightContext);
```
Comment (Hixie) [inline packages/flutter/lib/src/material/dropdown_menu.dart:527]: shouldn't we fix `Scrollable.ensureVisible`? Surely it's a bug that calling that doesn't work.

Comment (TahaTesser) [reply packages/flutter/lib/src/material/dropdown_menu.dart:527]: DropdownMenu` calls `Scrollable.ensureVisible` which then calls `ScrollPosition.ensureVisible`. However, after the `OverlayPortal` migration, it seems the correct scrollable context isn't passed in nested scroll views.

https://github.com/flutter/flutter/blob/1465da40b7c1899ce06fa55ff249ffc046182461/packages/flutter/lib/src/widgets/scrollable.dart#L471-L476

https://github.com/flutter/flutter/blob/1465da40b7c1899ce06fa55ff249ffc046182461/packages/flutter/lib/src/widgets/scrollable.dart#L1028

My fix finds the scrollable and calls `ScrollPosition.ensureVisible` directly. 

 I'm not expert in scrollables so I'll defer to @Piinks, if this makes sense.

Comment (TahaTesser) [reply packages/flutter/lib/src/material/dropdown_menu.dart:527]: I can reproduce the issue in the official `OverlayPortal` example by nesting scroll views. 

Hixie is right it is a bug in `Scrollable.ensureVisible`.  However, I'm not sure `ensureVisible` can access `ScrollableState` outside `OverlayPortal` context. 

### Code sample

<details>
<summary>expand to view the code sample</summary> 

```dart
import 'package:flutter/material.dart';

/// Flutter code sample for [OverlayPortal].

void main() => runApp(const OverlayPortalExampleApp());

class OverlayPortalExampleApp extends StatelessWidget {
  const OverlayPortalExampleApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(title: const Text('OverlayPortal Example')),
        body: const Center(child: ClickableTooltipWidget()),
      ),
    );
  }
}

class ClickableTooltipWidget extends StatefulWidget {
  const ClickableTooltipWidget({super.key});

  @override
  State<StatefulWidget> createState() => ClickableTooltipWidgetState();
}

class ClickableTooltipWidgetState extends State<ClickableTooltipWidget> {
  final OverlayPortalController _tooltipController = OverlayPortalController();

  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      child: Card(
        child: SingleChildScrollView(
          child: TextButton(
            onPressed: _tooltipController.toggle,
            child: DefaultTextStyle(
              style: DefaultTextStyle.of(context).style.copyWith(fontSize: 50),
              child: OverlayPortal(
                controller: _tooltipController,
                overlayChildBuilder: (BuildContext context) {
                  return  Positioned(
                    right: 50,
                    bottom: 50,
                    child: GestureDetector(
                      onTap: () {
                        print('tooltip tapped');
                        Scrollable.ensureVisible(context);
                      },
                      child: const ColoredBox(
                        color: Colors.amberAccent,
                        child: Text('tooltip'),
                      ),
                    ),
                  );
                },
                child: const Text('Press to show/hide tooltip'),
              ),
            ),
          ),
        ),
      ),
    );
  }
}
```

</details>

Comment (TahaTesser) [reply packages/flutter/lib/src/material/dropdown_menu.dart:527]: https://github.com/flutter/flutter/pull/67773 made similar fix for nested scroll views, it doesn't handle `OverlayPortal` scenario very well and this is where the error is coming.

I'll close this PR to find out why this fails in an `OverlayPortal` and fix it `ensureVisible` instead`.

**Comment by felixSchl** [comment]
Hey, I am running into this problem, too. How can I fix this in user-land or do I have to wait for an upstream fix?

**Comment by anthonyR012** [comment]
Hello, I would like to know if there is any news about this issue. It seems that I am facing the same error.

**Comment by TahaTesser** [comment]
This was fixed in https://github.com/flutter/flutter/pull/148897. It would be available in the next major stable release.

**Pull request closed without merging**